copy/progress_test.go

```diff
@@ -0,0 +1,112 @@
+package copy
+
+import (
+	"fmt"
+	"os/exec"
+	"testing"
+	"time"
+
+	"github.com/containers/image/docker/daemon"
+	"github.com/containers/image/signature"
+	"github.com/containers/image/types"
+	"github.com/docker/docker/reference"
+)
+
+// THIS TEST SHOULD BE SLOW. Read inner comments for details.
+func TestCopyProgress(t *testing.T) {
+	interval := 100 * time.Millisecond
+
+	// we shell out to avoid "client is newer than server" errors from docker
+	// which will happen when importing the client straight from github.
+	// This of course presumes docker is in the path.
+	if out, err := exec.Command("/bin/sh", "-c", "docker pull golang:latest").CombinedOutput(); err != nil {
+		t.Fatalf("%v: %s", err, out)
+	}
+
+	ref, err := daemon.ParseReference("docker.io/library/golang:latest")
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	img, err := ref.NewImage(nil)
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer img.Close()
+
+	tgtRef, err := reference.ParseNamed("docker.io/containers/test:latest")
+	if err != nil {
+		t.Fatal(err)
+	}
+	tgt, err := daemon.NewReference("", tgtRef)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	pc, err := signature.NewPolicyContext(&signature.Policy{
+		Default: []signature.PolicyRequirement{signature.NewPRInsecureAcceptAnything()},
+	})
+
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	signal := make(chan types.ProgressProperties)
+	errChan := make(chan error, 1)
+
+	// This validates that enough events arrived to satisfy the interval time. As
+	// a result it may require a little bit of work to get right before it will
+	// pass for all people all the time. Sorry in advance!
+	go func() {
+		var (
+			// if this is false, the time will be populated and this will be set to true.
+			// This is to ensure we're measuring the progress itself and not other
+			// operations.
+			timed bool
+			tm    time.Time
```

Comment:
So the variables really are `firstReportedTime` and `firstReportedTimeSet`? (It’s not that overlong variable names are desirable in principle, but in this case just naming the variables precisely would 1) eliminate the need for the comment, and 2) be explicit, when so many values in here are a “time”)

[May be irrelevant if this test is reworked with a special input stream.]